Add reference file automation scripts, and shape and symbol reference plugins

jedrichards opened the pull request:
Refs #525

mathieudutour commented:
Instead of creating 1 plugin for each file, perhaps it'd be easier to maintain 1 plugin with 1 command for each file? Also it'd be interesting to pretty print the JSON files

jedrichards commented:
@mathieudutour Agreed on both counts. Tasks for the next PR maybe?

jedrichards commented:
@mathieudutour Added some scripts to automate generating the files, and also added pretty printing the JSON and moved to a single plugin with multiple commands, as suggested. See what you think ...

mathieudutour commented on scripts/generate-reference-files-md.js:
```diff
@@ -0,0 +1,58 @@
+const pages = resolve(__dirname, '../docs/pages')
+const pluginPath = resolve(dir, './plugin.sketchplugin')
+const manifest = require(resolve(pluginPath, 'Contents/Sketch/manifest.json'))
+const repoPath =
```
I think you can take `https://github.com/BohemianCoding/SketchAPI` from the package.json

jedrichards replied:
Yep.

jedrichards replied:
Hmm, the `respository.url` field in package.json is for urls to the actual `.git` repo, not the http location of the GitHub web ui. Going to add the `homepage` field to package.json, and grab from there ...

mathieudutour commented on reference-files/plugin.sketchplugin/Contents/Sketch/shapes.js:
```diff
@@ -26,7 +26,7 @@ function main(ctx) {
+  doc.save(String(ctx.savePath), {
     saveMode: Document.SaveMode.SaveAs,
   })
   doc.close()
```
I missed that earlier but the `.close()` should be in the callback of the `.save` method

jedrichards replied:
Ah ok, yes makes sense.

jedrichards commented on scripts/generate-reference-files-md.js:
```diff
@@ -0,0 +1,58 @@
+  const mdPath = resolve(pages, `file-format-reference-${cmd.identifier}.md`)
+
+  writeFileSync(mdPath, md, 'utf8')
+  execSync(`npm run prettier -- --write "${mdPath}"`, { encoding: 'utf8' })
```
Ah of course, `npm run` will pick up prettify automatically, good catch.

jedrichards commented on scripts/generate-reference-files.js:
```diff
@@ -0,0 +1,63 @@
+const del = require('del')
+const { resolve } = require('path')
+
+const sketchVersion = process.argv[2] || '55'
```
Ok roger.

jedrichards replied:
The `sketchtool` output isn't exactly nicely machine readable:

```
sketchtool Version 55.1 (78136)
```

Easy enough to parse out though, unless there's another way of getting the Sketch version in a script?

jedrichards commented:
> One thing tho is that you are using sketchtool directly

Not too concerned about this at the moment, since I mention it in the script comments, and I'm guessing it'll mainly be me running the script for the time being! Your approach of passing in the path sounds good as a future improvement though

> Another thing: it'd be nice to change the UUID of the generated files to a constant string so that diffs are easier to read.

What would be the approach here? Manipulate the generated JSON itself, or is there a way to use specific/hard coded uuids through the API? I'm not sure I'd like to add too much code to the script that was dependant on particular a data structures/field name in the JSON if I can get away with it ...

mathieudutour commented:
> What would be the approach here?

I was thinking about a simple regex to run on the JSON file: `/"do_objectID": "([A-D0-9\-]+)"/`. There is no support for a hardcoded uuid in the API because it's usually a bad idea to change the id of sketch objects (Sketch is going to be really confused if you change certain IDs)
